File: algorithms-week7/connected_components/python/connected_components.py
#Uses python3

import sys


def number_of_components(adj):
    result = 0
    n = len(adj)
    visited = [0] * n

    for v in range(n):
        if visited[v] == 0:
            dfs(adj, visited, v)
            result += 1

    return result

def dfs(adj, visited, v):
    visited[v] = 1

    for w in adj[v]:
        if visited[w] == 0:
            dfs(adj, visited, w)

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(number_of_components(adj))

File: algorithms-week7/connected_components/java/src/ConnectedComponents.java
import java.util.ArrayList;
import java.util.Scanner;

public class ConnectedComponents {
    private static int numberOfComponents(ArrayList<Integer>[] adj) {
        int result = 0;
        int n = adj.length;
        int[] visited = new int[n];

        for (int i = 0; i < n; i++) {
            if (visited[i] == 0) {
                dfs(adj, visited, i);
                result++;
            }
        }
        return result;
    }

    private static void dfs(ArrayList<Integer>[] adj, int[] visited, int v) {
        visited[v] = 1;

        for (int w : adj[v]) {
            if (visited[w] == 0) {
                dfs(adj, visited, w);
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        System.out.println(numberOfComponents(adj));
    }
}

